Add Reset Of User Password From Admin

GuenoleCHERUBIN opened the pull request:
| Q             | A
| ------------- | ---
| Is bugfix?    | No
| New feature?  | Yes

Added the possibility for an admin to reset
an user's password.
It sends an email to the user, with a token, to create
a new password.

GuenoleCHERUBIN committed:
Add Reset Of User Password From Admin

Added the possibility for an admin to reset
an user's password.
It sends an email to the user, with a token, to create
a new password.

EmileCalixte commented on views/management/accounts.php:
```diff
@@ -59,8 +59,17 @@
+                /** @var \app\models\databaseModels\User $user */
 
-
+                return "<a href='/management/reset-password/" . $user->id . "'  id='reset-password'>Reset User Password<a/>";
```
CSRF : NEVER use a link (GET method) to EDIT the application state. Do a POST form

EmileCalixte commented on controllers/ManagementController.php:
```diff
@@ -330,5 +330,67 @@ private function createUserToken()
+     * @throws CannotSaveException
+     */
+    public function actionResetPassword($id = null)
+    {
```
CSRF : At the beginning of your function, check if the request is POST by throwing a `MethodNotAllowedHttpException` if `\Yii::$app->request->isPost` is `false`

EmileCalixte commented on controllers/ManagementController.php:
```diff
@@ -330,5 +330,67 @@ private function createUserToken()
+     */
+    private function resetPassword($user)
+    {
+        $user->password_hash = null;
```
Don't forget to set `$user->is_active` to `false`, otherwise it will cause a fatal error when the user tries to log in without re-creating his password.

GuenoleCHERUBIN committed:
Changed The Way Reset Password Is Done So It's Safer

Changed the way the admin reset the password so it's safer
and more logical.
Now set is_active to false when admin reset the password,
and set is-active to true when the user recreate is password.

EmileCalixte commented on models/forms/ResetPasswordForm.php:
```diff
@@ -0,0 +1,21 @@
+            [['id'], 'required'],
+            // id must be an integer
+            ['id', 'integer'],
+            ];
```
Indentation error 👀

EmileCalixte commented on views/management/accounts.php:
```diff
@@ -59,8 +60,26 @@
+                </form>
+                <?php
+                return ob_get_clean();
+                //return "<a href='/management/reset-password/" . $user->id . "'  id='reset-password'>Reset User Password<a/>";
```
You can remove this comment

GuenoleCHERUBIN committed:
Ghanged LoginForm So That If Manage Null Password

We don't desactivate an account on password reset nor
reactivate it if the user reacreate a new password.
If the password for the email given is null,
show an error.

GuenoleCHERUBIN committed:
Changes After Comments

EmileCalixte commented on models/forms/LoginForm.php:
```diff
@@ -46,7 +46,7 @@ public function validatePassword($attribute, $params)
             $user = $this->getUser();
 
-            if (!$user || !$user->validatePassword($this->password)) {
+            if (!$user || !$user->validatePassword($this->password) || $user->password_hash === null) {
```
You must check if the password is null before validating it, otherwise it's useless...
`if (!$user || $user->password_hash === null || !$user->validatePassword($this->password))`

GuenoleCHERUBIN committed:
Change After Comments

Check if a post is given to reset-password action.
Swapped conditions order in LoginForm, was creating an error.

EmileCalixte committed:
Merge branch 'master' into reset-password-admin
